README.md
# Number Place

[![MIT License](https://img.shields.io/github/license/AlbinoGeek/number-place.svg)](https://github.com/AlbinoGeek/number-place/blob/master/LICENSE)
[![Go Version](https://img.shields.io/github/go-mod/go-version/AlbinoGeek/number-place.svg)](https://github.com/AlbinoGeek/number-place)
[![Maintainability](https://api.codeclimate.com/v1/badges/be0523753694eee85927/maintainability)](https://codeclimate.com/github/AlbinoGeek/number-place/maintainability)  
[![CI](https://github.com/AlbinoGeek/number-place/workflows/CI/badge.svg?branch=main)](#)
[![GoReportCard](https://goreportcard.com/badge/github.com/AlbinoGeek/number-place)](https://goreportcard.com/report/github.com/AlbinoGeek/number-place)
[![Coverage Status](https://coveralls.io/repos/github/AlbinoGeek/number-place/badge.svg?branch=develop)](https://coveralls.io/github/AlbinoGeek/number-place?branch=develop)

[![](cmd/testdata/start.png)](cmd/testdata/start.png)

## Features

Share puzzles simply by copying a string:

```
3,3,3,3,53-6---98-7-195----------6-8--4--7---6-8-3-2---3--1--6-6----------419-8-28---5-79
```

Cells:

- Click/Tap Toggle Select
- Drag to Select Multiple
- Unlimited Undo Levels

Help:

- Highlights Simple Mistakes

### Keyboard Controls

With Cells Selected:

- Use the alpha keys or numpad to set the center value
- Use the `del`ete or `backspace` keys to clear all values

### Puzzles Supported

- [X] Sudoku
  - [X] Standard Grid (9x9, 3x3 boxes)
    - [X] Classic
    - [ ] Quadratum latinum (uses roman numerals for values)
    - [ ] Alphabetical (uses a key of letters for values)
  - [ ] Mini Grid (6x6, 3x2 boxes)
  - [ ] Giant Grid (16x16, 4x4 boxes)
- [ ] Constrainted-Based Gridded Sudoku:
  - [ ] Standard Grid (9x9, 3x3 boxes)
    - [ ] "Killer"
    - [ ] Greater Than
    - [ ] XV

If you know of a puzzle type not listed above, please [request it!](https://github.com/AlbinoGeek/number-place/issues/new?assignees=AlbinoGeek&labels=enhancement&template=feature-request.md&title=%5BFEATURE+REQUEST%5D)

## Usage

[Download a Release](https://github.com/AlbinoGeek/number-place/releases) or [Get the Sources](#building-from-source)

## Building From Source

1. Install Golang
1. Clone the repository
1. Run `make`

```bash
# Install Golang [apt / yum / dnf install golang]
$ git clone https://github.com/AlbinoGeek/number-place
$ cd number-place
$ make all
```

The binary will be built into the `_dist` directory.

## License

This project is licensed under the terms of the [MIT License](/LICENSE).


CHANGELOG.md
# Changelog

## Unreleased

**Added**

- Button allows you to undo any value set.
- Simple mistakes are highlighted as you play. (#4)
- Button to check for mistakes. (#4)
- Set/Clear values with alpha keys/numpad. (#5)
- Button to reset puzzle to initial state. (#20)
- New Game Timer shown in the bottom right, stops when puzzle solved. (#27)

**Changed**

- More uniform Focus/Hover/Selected colours.

**Fixed**

- Many slowdowns in check methods. (#11)
- Bug where boxes showed cells in the wrong order in a different orientation. (#21, #22)
- Puzzle Check button now also tells you if you have found a possible solution. (#27)

## v0.1

- Initial Tagged Build
- Support for Classic Sodoku (9x9).
- Drag-select multiple boxes.


cmd/main.go
package main

import (
	"strconv"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/dialog"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"
)

//go:generate go run gen.go

// TODO: These become configuration

// HighlightMistakes will become a configuration variable representing whether
// summary mistakes are checked for with every set.
var HighlightMistakes = true

var wikipedia = `3,3,3,3,53-6---98-7-195----------6-8--4--7---6-8-3-2---3--1--6-6----------419-8-28---5-79`

func main() {
	a := app.NewWithID("com.github.albinogeek.number-place")
	b, _ := start(a)
	b.load(wikipedia)
	b.gameTimer.Show()
	b.gameTimer.Refresh()
	a.Run()
}

func start(a fyne.App) (*board, fyne.Window) {
	w := a.NewWindow("Number Place")
	b := newBoard(3, 3, 3, 3)

	uiInit(b, w)

	w.Show()

	return b, w
}

func uiInit(b *board, w fyne.Window) {
	controls := make([]fyne.CanvasObject, 0)

	values := make(map[string]bool)
	// TODO: Support other digit systems (such as HEX for sandwiche or Giant)
	for i := 0; i < b.boxWidth*b.boxesWide; i++ {
		v := strconv.Itoa(1 + i)
		controls = append(controls, widget.NewButton(v, setSelected(b, v)))
		values[v] = true
	}

	w.Canvas().SetOnTypedKey(keyHandler(b, values))

	controlArea := container.NewPadded(container.NewVBox(
		fyne.NewContainerWithLayout(
			layout.NewGridLayout(b.boxWidth),
			controls...,
		),
		widget.NewSeparator(),
		fyne.NewContainerWithLayout(
			layout.NewGridLayout(3),
			widget.NewButtonWithIcon("", theme.CancelIcon(), clearSelected(b)),
			widget.NewButtonWithIcon("", theme.ContentUndoIcon(), b.Undo),
			widget.NewButtonWithIcon("", theme.ConfirmIcon(), func() {
				if err := b.check(); err != nil {
					dialog.ShowError(err, w)
					return
				}

				dialog.ShowInformation("Check Passed",
					"I don't see any mistakes right now.\nIt's up to you to complete the puzzle.", w)
			}),
		),
		widget.NewSeparator(),
		widget.NewButtonWithIcon("", theme.DeleteIcon(), b.Reset),
		layout.NewSpacer(),
		container.NewHBox(b.gameTimer),
	))

	w.SetContent(container.NewBorder(
		nil, nil, nil,

		// Right
		controlArea,

		// Objects
		b.Container,
	))

	w.SetFixedSize(true)
	w.CenterOnScreen()
}

func keyHandler(b *board, values map[string]bool) func(*fyne.KeyEvent) {
	key := func(value string) {
		if yes, ok := values[value]; !ok || !yes {
			return
		}

		setSelected(b, value)()
	}

	return func(ke *fyne.KeyEvent) {
		switch ke.Name {
		case fyne.Key1:
			key("1")
		case fyne.Key2:
			key("2")
		case fyne.Key3:
			key("3")
		case fyne.Key4:
			key("4")
		case fyne.Key5:
			key("5")
		case fyne.Key6:
			key("6")
		case fyne.Key7:
			key("7")
		case fyne.Key8:
			key("8")
		case fyne.Key9:
			key("9")
		case fyne.KeyDelete:
			clearSelected(b)()
		case fyne.KeyBackspace:
			clearSelected(b)()
		}
	}
}

func clearSelected(b *board) func() {
	return func() {
		setSelected(b, "")()
	}
}

func setSelected(b *board, value string) func() {
	return func() {
		b.registerUndo()

		b.mu.Lock()
		for _, c := range b.cells {
			if c.selected {
				c.selected = false
				c.SetCenter(value)
			} else {
				c.SetMistake(false)
			}
		}
		b.mu.Unlock()

		b.check()
	}
}
